File: src/R/fitnessFunctions.R
ulibrary(dplyr)
library(tidyverse)
library(purrr)

quadratic <- function(population) {
  reduce(population, function(cum, x){cum+x*x}, .init=0)
}

elliptic <- function(population) {
  D = ncol(population)
  reduce(population, function(cum, x){
    list(cum[[1]] + 10^(6*(cum[[2]]-1)/(D-1))*x*x, cum[[2]]+1)
  }, .init=list(0, 1))[[1]]
}

rastrigin <- function(population) {
    reduce(population, function(cum, x){
        cum + x*x - 10*cos(2*pi*x)+10
    }, .init=0)
}

rosenbrock <- function(population) {
    D <- ncol(population)
    reduce2(population[,1:D-1], population[,2:D], function(cum, x, y){
        cum + 100*(x*x - y)^2 + (x-1)^2
    }, .init=0)
}

ackley <- function(population) {
    quad <- quadratic(population)
    cosfunc <- reduce(population, function(cum, x){
        cum + cos(2*pi*x)
    }, .init=0)
    D <- ncol(population)
    reduce(population, function(cum, x){
        cum - 20*exp(-0.2*sqrt(quad/D))-exp(cosfunc/D)+20+exp(1)
    }, .init=0)
}

File: src/py/fitnessFunctions.py
from functools import reduce
from math import pi, cos, exp, sqrt


def spherical_pop(population):
    return list(map(spherical, population))


def elliptic_pop(population):
    return list(map(elliptic, population))


def rastrigin_pop(population):
    return list(map(rastrigin, population))


def rosenbrock_pop(population):
    return list(map(rosenbrock, population))


def ackley_pop(population):
    return list(map(ackley, population))


def spherical(point):
    return reduce(lambda cum, x: cum + x * x, point, 0)


def elliptic(point):
    D = len(point)
    return reduce(lambda cum, x: (cum[0] + pow(10, 6 * (cum[1] - 1) / (D - 1)) * x * x, cum[1] + 1), point,
                  (0, 1))[0]


def rastrigin(point):
    return reduce(lambda cum, x: cum + x * x - 10 * cos(2 * pi * x) + 10, point, 0)


def rosenbrock(point):
    D = len(point)
    return reduce(lambda cum, p: (lambda x, y: cum + 100 * pow(x * x - y, 2) + pow(x - 1, 2))(*p),
                  zip(point[:D - 1], point[1:]), 0)


def ackley(point):
    quad = spherical(point)
    cosfunc = reduce(lambda cum, x: cum + cos(2 * pi * x), point, 0)
    D = len(point)
    return reduce(lambda cum, x: cum - 20 * exp(-0.2 * sqrt(quad / D)) - exp(cosfunc / D) + 20 + exp(1), point, 0)
